src/vm/eventpipe.cpp

```diff
@@ -515,15 +524,51 @@ void QCALLTYPE EventPipeInternal::DeleteProvider(
     END_QCALL;
 }
 
+struct EventProviderEventData
+{
+    unsigned long Ptr;
+    unsigned int Size;
+    unsigned int Reserved;
+};
+
 void QCALLTYPE EventPipeInternal::WriteEvent(
     INT_PTR eventHandle,
+    unsigned int eventID,
     void *pData,
-    unsigned int length)
+    unsigned int dataCount)
 {
     QCALL_CONTRACT;
     BEGIN_QCALL;
 
-    // TODO
+    _ASSERTE(eventHandle != NULL);
+    EventPipeEvent *pEvent = reinterpret_cast<EventPipeEvent *>(eventHandle);
+
+    if (pData == NULL)
+    {
+        EventPipe::WriteEvent(*pEvent, NULL, 0);
+    }
+    else
+    {
+        // pData is an array of pointers to blobs with Ptr of objects and Size of objects
+        // and we merge them here
+        EventProviderEventData *pNativeData = (EventProviderEventData *)pData;
+        unsigned int length = 0;
+        for (int i = 0; i < dataCount; i++)
+        {
+            length += pNativeData->Size;
+        }
+
+        BYTE *buffer = new BYTE[length];
```

Comment:
Consider: A static buffer of say 100 bytes.  If the data size is greater, then allocate.  I suspect many events will be under 100 bytes and then you won't have to alloc and free on the heap.